docs/src/proposals/blockstore-rocksdb-compaction.md

```diff
@@ -0,0 +1,234 @@
+and the read amplification is the same as the number of SST files.
+
+Luckily, it is not the case in Solana's BlockStore workload because write-keys
+are mostly monotonically increasing over time.  Thus, the resulting SST files
+will mostly have disjoint key range.  For most reads, it can be answered by
+hitting mostly one SST file or at most two SST files in case there're some
+small out-of-order writes.
+
+To sum up, if no other manual compaction is issued for quickly picking up
+deletions, FIFO Compaction offers the following amplification factors
+in Solana's BlockStore use case:
+
+- Write Amplification: 1 (all data is written once without compaction.)
+- Read Amplification: < 1.1 (assuming each SST file has 10% overlapping key
+  range with another SST file.)
+- Space Amplification: 1 (same data never be written in more than one SST file,
+  and no additional temporary space required for compaction.)
+
+### Compaction Style Migration
+As FIFO Compaction assumes each SST file represents one logical level, RocksDB
+instances compacted by other compaction styles are valid FIFO compacted
+instances.  Migration from Level Compaction to FIFO Compaction should be simple.
+
+### Other Detailed Settings
+- FIFO Compaction requires all SST files to be always opened.  On the RocksDB
+  side, max_open_files must be set to -1 (meaning unlimited.)
+
+### Benefits
+#### No Write Stalls: Writes Never Wait for Compactions
+Because data is most written in sorted order in Solana's BlockStore workload,
+each read can be answered by hitting mostly one file (and in the boundary case
+where two SST files with some small key-range overlap.)  In other words there
+is no need to run compactions in order to maintain read amplification.  As a
+result, rate-limiting or stalls will never happen to writes in order to wait
+for compactions which try to merge logical levels.
+
+#### Minimize Disk I/Os by avoiding unnecessary compactions
+As mentioned in the opportunities section, our SST files are naturally range
+partitioned, and there will be only one entry associated with same key.  This
+two facts together means we can avoid all the background compactions that
+tries to either merge entries or perform range partition.  As a result, the
+majority of the I/Os contributed by compactions can be avoided.
+
+### Potential Limitations and Possible Solutions
+While FIFO Compaction offers great amplification factors without write stalls
+which is suitable for Solana's BlockStore, here I listed its limitations and
+possible solutions in case we want to further optimize certain scenarios in
+the future.
+
+#### No Direct Support for Background Compactions by Default
+FIFO Compaction doesn't support periodic background compactions by default.
+As a result, in case there're situations where we want to compact a SST file
+which is not old enough to reach the configured TTL such as to diguest certain
+deletion entries, this must be done via running manual compactions.
```

Comment:
What would this look like for writes that are "updates" to metadata fields?

Using the `INDEX_CF` as an example, `INDEX_CF` contains a `BTreeSet<u64>` for received data and coding shreds. Supposing all of our `blockstore::insert_shreds()` calls for this slot have a small number of shreds per insert, we might have 100's (or maybe even > 1000) of updates for a slot. As the slot fills up, the size of the `BTreeSet<u64>` becomes non-trivial:

500 shreds * 8 bytes / shred index * 2 (factoring data and coding) = 8000 bytes (ignoring the overhead of datastructure)

Reply:
Thanks for pointing this out.  I think I didn't make it clear.

The FIFO Compaction should only be used for the column family that stores the actual shred data.  Metadata stored in different column families should still use level compaction.